go/vt/vttablet/tabletmanager/vreplication/vstreamer_client.go

```diff
@@ -0,0 +1,213 @@
+	VStream(ctx context.Context, startPos string, filter *binlogdatapb.Filter, send func([]*binlogdatapb.VEvent) error) error
+
+	// VStreamRows streams rows of a table from the specified starting point.
+	VStreamRows(ctx context.Context, query string, lastpk *querypb.QueryResult, send func(*binlogdatapb.VStreamRowsResponse) error) error
+}
+
+// TabletVStreamerClient a vstream client backed by vttablet
+type TabletVStreamerClient struct {
+	// mu protects isOpen, streamers, streamIdx and kschema.
+	mu sync.Mutex
+
+	isOpen bool
+
+	tablet         *topodatapb.Tablet
+	target         *querypb.Target
+	tsQueryService queryservice.QueryService
+}
+
+// MySQLVStreamerClient a vstream client backed by MySQL
+type MySQLVStreamerClient struct {
+	// mu protects isOpen, streamers, streamIdx and kschema.
+	mu sync.Mutex
+
+	isOpen bool
+
+	sourceConnParams *mysql.ConnParams
+	vsEngine         *vstreamer.Engine
+}
+
+// NewTabletVStreamerClient creates a new TabletVStreamerClient
+func NewTabletVStreamerClient(tablet *topodatapb.Tablet) *TabletVStreamerClient {
+	return &TabletVStreamerClient{
+		tablet: tablet,
+		target: &querypb.Target{
+			Keyspace:   tablet.Keyspace,
+			Shard:      tablet.Shard,
+			TabletType: tablet.Type,
+		},
+	}
+}
+
+// Open part of the VStreamerClient interface
+func (vsClient *TabletVStreamerClient) Open(ctx context.Context) (err error) {
+	vsClient.mu.Lock()
+	defer vsClient.mu.Unlock()
+	if vsClient.isOpen {
+		return nil
+	}
+	vsClient.isOpen = true
+
+	vsClient.tsQueryService, err = tabletconn.GetDialer()(vsClient.tablet, grpcclient.FailFast(false))
+	return err
+}
+
+// Close part of the VStreamerClient interface
+func (vsClient *TabletVStreamerClient) Close(ctx context.Context) (err error) {
+	if !vsClient.isOpen {
+		return nil
+	}
+	return vsClient.tsQueryService.Close(ctx)
+}
+
+// VStream part of the VStreamerClient interface
+func (vsClient *TabletVStreamerClient) VStream(ctx context.Context, startPos string, filter *binlogdatapb.Filter, send func([]*binlogdatapb.VEvent) error) error {
+	if !vsClient.isOpen {
+		return errors.New("Can't VStream without opening client")
+	}
+	return vsClient.tsQueryService.VStream(ctx, vsClient.target, startPos, filter, send)
+}
+
+// VStreamRows part of the VStreamerClient interface
+func (vsClient *TabletVStreamerClient) VStreamRows(ctx context.Context, query string, lastpk *querypb.QueryResult, send func(*binlogdatapb.VStreamRowsResponse) error) error {
+	if !vsClient.isOpen {
+		return errors.New("Can't VStreamRows without opening client")
+	}
+	return vsClient.tsQueryService.VStreamRows(ctx, vsClient.target, query, lastpk, send)
+}
+
+// NewMySQLVStreamerClient is a vstream client that allows you to stream directly from MySQL.
+// In order to achieve this, the following creates a vstreamer Engine with a dummy in memorytopo.
+func NewMySQLVStreamerClient(sourceConnParams *mysql.ConnParams) *MySQLVStreamerClient {
```

Comment:
I'm thinking this function can pull the dbconfigs based on the external repl user name. Then you don't have to pass it through to vreplication.Engine.

Reply:
My thinking here is that the end goal is to be able to point to any external DB, having a this parameter here will make it more flexible. 

I think we shouldn't rely heavily in the repl username as we would like to refactor that soon. 

What do you think?